src/main/java/org/apache/lucene/BoundedLinearFeatureQuery.java

```diff
@@ -0,0 +1,235 @@
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * The feature queries of input tokens are wrapped by lucene BooleanQuery, which use WAND algorithm
+ * to accelerate the execution. The WAND algorithm leverage the score upper bound of sub-queries to
+ * skip non-competitive tokens. However, origin lucene FeatureQuery use Float.MAX_VALUE as the score
+ * upper bound, and this invalidates WAND.
+ *
+ * To mitigate this issue, we rewrite the FeatureQuery to BoundedLinearFeatureQuery. The caller can
+ * set the token score upperbound of this query. And according to our use case, we use LinearFunction
+ * as the score function.
+ *
+ * This class combines both <a href="https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/document/FeatureQuery.java">FeatureQuery</a>
+ * and <a href="https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/document/FeatureField.java">FeatureField</a> together
+ * and will be deprecated after OpenSearch upgraded lucene to version 9.8.
+ */
+
+public final class BoundedLinearFeatureQuery extends Query {
+
+    private final String fieldName;
+    private final String featureName;
+    private final Float scoreUpperBound;
+
+    public BoundedLinearFeatureQuery(String fieldName, String featureName, Float scoreUpperBound) {
+        this.fieldName = Objects.requireNonNull(fieldName);
+        this.featureName = Objects.requireNonNull(featureName);
+        this.scoreUpperBound = Objects.requireNonNull(scoreUpperBound);
+    }
+
+    @Override
+    public Query rewrite(IndexSearcher indexSearcher) throws IOException {
+        // LinearFunction return same object for rewrite
+        return super.rewrite(indexSearcher);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        BoundedLinearFeatureQuery that = (BoundedLinearFeatureQuery) obj;
+        return Objects.equals(fieldName, that.fieldName)
+            && Objects.equals(featureName, that.featureName)
+            && Objects.equals(scoreUpperBound, that.scoreUpperBound);
+    }
+
+    @Override
+    public int hashCode() {
+        int h = getClass().hashCode();
+        h = 31 * h + fieldName.hashCode();
+        h = 31 * h + featureName.hashCode();
+        h = 31 * h + scoreUpperBound.hashCode();
+        return h;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        if (!scoreMode.needsScores()) {
+            // We don't need scores (e.g. for faceting), and since features are stored as terms,
+            // allow TermQuery to optimize in this case
+            TermQuery tq = new TermQuery(new Term(fieldName, featureName));
+            return searcher.rewrite(tq).createWeight(searcher, scoreMode, boost);
+        }
+
+        return new Weight(this) {
```

Comment:
can we move this weight class as inner class or a separate class?

Reply:
This is copied from lucene FeatureQuery, I think we can save the UT effort since this feature will only live a very short time(one version) and the added method is already covered.